File: examples/canonical-faker/scala/src/main/scala/io/agentcontextbase/faker/Pipeline.scala
package io.agentcontextbase.faker

import net.datafaker.Faker
import java.time.format.DateTimeFormatter
import java.time.{Instant, ZoneOffset, ZoneId}
import java.util.concurrent.TimeUnit
import scala.util.Random

object Pipeline {
  private val formatter = DateTimeFormatter.ISO_INSTANT

  def generate(profile: Profile): Dataset = {
    val rng = new Random(profile.seed)
    val faker = new Faker(new java.util.Random(profile.seed))

    // ZIO Stream could replace LazyList for medium/large profiles
    // when backpressure and async chunked writes matter.

    val orgs = buildOrgs(profile.numOrgs, faker, rng)
    val users = buildUsers(profile.numUsers, faker, rng)

    val (memberships, orgMemberMap) = buildMemberships(orgs, users, rng)
    val (projects, projectOrgMap) = buildProjects(orgs, orgMemberMap, faker, rng)
    val apiKeys = buildApiKeys(orgs, orgMemberMap, faker, rng)
    val invitations = buildInvitations(orgs, orgMemberMap, users, faker, rng)
    val auditEvents = buildAuditEvents(projects, orgMemberMap, projectOrgMap, apiKeys, invitations, rng)

    val dataset = Dataset(
      organizations = orgs.toVector,
      users = users.toVector,
      memberships = memberships.toVector,
      projects = projects.toVector,
      audit_events = auditEvents.toVector,
      api_keys = apiKeys.toVector,
      invitations = invitations.toVector
    )

    val report = Validate.check(dataset).copy(seed = profile.seed, profileName = profile.name)
    if (!report.ok) {
      throw new RuntimeException(s"Validation failed: ${report.violations.mkString(", ")}")
    }

    dataset.copy(report = Some(report))
  }

  private def buildOrgs(count: Int, faker: Faker, rng: Random): LazyList[Organization] = {
    var seenSlugs = Set.empty[String]
    var seenEmails = Set.empty[String]

    LazyList.continually {
      val name = faker.company().name()
      val baseSlug = name.toLowerCase.replaceAll("[^a-z0-9]+", "-").stripMargin('-')
      var slug = baseSlug
      var suffix = 1
      while (seenSlugs.contains(slug)) {
        slug = s"$baseSlug-$suffix"
        suffix += 1
      }
      seenSlugs += slug

      var email = faker.internet().emailAddress()
      while (seenEmails.contains(email)) {
        email = faker.internet().emailAddress()
      }
      seenEmails += email

      val plan = Distributions.weightedPlan(rng)
      val region = Distributions.weightedRegion(rng)
      
      val baseDate = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
      val range = 3L * 365 * 24 * 60 * 60 * 1000
      val createdAtInstant = baseDate - (rng.nextDouble() * range).toLong
      val createdAt = Instant.ofEpochMilli(createdAtInstant).toString

      Organization(
        id = Pools.deterministicUuid(rng),
        name = name,
        slug = slug,
        plan = plan,
        region = region,
        created_at = createdAt,
        owner_email = email
      )
    }.take(count)
  }

  private def buildUsers(count: Int, faker: Faker, rng: Random): LazyList[User] = {
    var seenEmails = Set.empty[String]
    LazyList.continually {
      var email = faker.internet().emailAddress()
      while (seenEmails.contains(email)) {
        email = faker.internet().emailAddress()
      }
      seenEmails += email

      val fullName = faker.name().fullName()
      val locale = if (rng.nextDouble() < 0.6) "en-US" else "en-GB"
      val timezone = if (locale == "en-US") "America/New_York" else "Europe/London"
      
      val baseDate = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
      val range = 3L * 365 * 24 * 60 * 60 * 1000
      val createdAtInstant = baseDate - (rng.nextDouble() * range).toLong
      val createdAt = Instant.ofEpochMilli(createdAtInstant).toString

      User(
        id = Pools.deterministicUuid(rng),
        email = email,
        full_name = fullName,
        locale = locale,
        timezone = timezone,
        created_at = createdAt
      )
    }.take(count)
  }

  private def buildMemberships(
    orgs: LazyList[Organization],
    users: LazyList[User],
    rng: Random
  ): (LazyList[Membership], Map[String, Vector[String]]) = {
    val allUserIds = users.map(_.id).toVector
    var memberships = Vector.empty[Membership]
    var orgMemberMap = Map.empty[String, Vector[String]]

    orgs.foreach { org =>
      val numMembers = 3 + rng.nextInt(48) // 3 to 50
      val memberIds = rng.shuffle(allUserIds).take(numMembers)
      orgMemberMap += (org.id -> memberIds)

      val orgMemberships = memberIds.zipWithIndex.map { case (userId, idx) =>
        val role = if (idx == 0) "owner" else Distributions.weightedRole(rng)
        val invitedBy = if (idx == 0) "" else memberIds(0) // first is owner

        val orgCreatedInstant = Instant.parse(org.created_at).toEpochMilli
        val now = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
        val range = math.max(1000L, now - orgCreatedInstant)
        val joinedAtInstant = orgCreatedInstant + (rng.nextDouble() * range).toLong
        val joinedAt = Instant.ofEpochMilli(joinedAtInstant).toString

        Membership(
          id = Pools.deterministicUuid(rng),
          org_id = org.id,
          user_id = userId,
          role = role,
          joined_at = joinedAt,
          invited_by = invitedBy
        )
      }
      memberships ++= orgMemberships
    }

    (memberships.to(LazyList), orgMemberMap)
  }

  private def buildProjects(
    orgs: LazyList[Organization],
    orgMemberMap: Map[String, Vector[String]],
    faker: Faker,
    rng: Random
  ): (LazyList[Project], Map[String, String]) = {
    var projects = Vector.empty[Project]
    var projectOrgMap = Map.empty[String, String]

    orgs.foreach { org =>
      val numProjects = 1 + rng.nextInt(20) // 1 to 20
      val memberIds = orgMemberMap(org.id)

      val orgProjects = (1 to numProjects).map { _ =>
        val createdBy = memberIds(rng.nextInt(memberIds.length))
        val status = Distributions.weightedProjectStatus(rng)

        val orgCreatedInstant = Instant.parse(org.created_at).toEpochMilli
        val now = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
        val range = math.max(1000L, now - orgCreatedInstant)
        val createdAtInstant = orgCreatedInstant + (rng.nextDouble() * range).toLong
        val createdAt = Instant.ofEpochMilli(createdAtInstant).toString

        val p = Project(
          id = Pools.deterministicUuid(rng),
          org_id = org.id,
          name = faker.app().name(),
          status = status,
          created_by = createdBy,
          created_at = createdAt
        )
        projectOrgMap += (p.id -> org.id)
        p
      }
      projects ++= orgProjects
    }

    (projects.to(LazyList), projectOrgMap)
  }

  private def buildApiKeys(
    orgs: LazyList[Organization],
    orgMemberMap: Map[String, Vector[String]],
    faker: Faker,
    rng: Random
  ): LazyList[ApiKey] = {
    orgs.flatMap { org =>
      val numKeys = rng.nextInt(11) // 0 to 10
      val memberIds = orgMemberMap(org.id)
      if (memberIds.isEmpty) Seq.empty
      else {
        (1 to numKeys).map { _ =>
          val createdBy = memberIds(rng.nextInt(memberIds.length))

          val orgCreatedInstant = Instant.parse(org.created_at).toEpochMilli
          val now = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
          val range = math.max(1000L, now - orgCreatedInstant)
          val createdAtInstant = orgCreatedInstant + (rng.nextDouble() * range).toLong
          val createdAt = Instant.ofEpochMilli(createdAtInstant).toString
          
          val lastUsedAtInstant = createdAtInstant + (rng.nextDouble() * (now - createdAtInstant)).toLong
          val lastUsedAt = if (rng.nextBoolean()) Instant.ofEpochMilli(lastUsedAtInstant).toString else null

          ApiKey(
            id = Pools.deterministicUuid(rng),
            org_id = org.id,
            created_by = createdBy,
            name = faker.lorem().word(),
            key_prefix = s"tc_${faker.internet().password(8, 8, true, true, true)}",
            created_at = createdAt,
            last_used_at = lastUsedAt
          )
        }
      }
    }
  }

  private def buildInvitations(
    orgs: LazyList[Organization],
    orgMemberMap: Map[String, Vector[String]],
    users: LazyList[User],
    faker: Faker,
    rng: Random
  ): LazyList[Invitation] = {
    val allEmails = users.map(_.email).toSet
    orgs.flatMap { org =>
      val numInvs = rng.nextInt(6) // 0 to 5
      val memberIds = orgMemberMap(org.id)
      if (memberIds.isEmpty) Seq.empty
      else {
        (1 to numInvs).map { _ =>
          val invitedBy = memberIds(rng.nextInt(memberIds.length))
          
          var email = faker.internet().emailAddress()
          while (allEmails.contains(email)) {
            email = faker.internet().emailAddress()
          }

          val baseDateStr = "2024-01-01T00:00:00Z"
          val baseDateInstant = Instant.parse(baseDateStr)
          val expiresAt = baseDateInstant.plusSeconds(rng.nextInt(30) * 86400L).toString
          val acceptedAt = if (rng.nextDouble() < 0.4) baseDateInstant.minusSeconds(rng.nextInt(30) * 86400L).toString else null

          Invitation(
            id = Pools.deterministicUuid(rng),
            org_id = org.id,
            invited_email = email,
            role = Distributions.weightedRole(rng),
            invited_by = invitedBy,
            expires_at = expiresAt,
            accepted_at = acceptedAt
          )
        }
      }
    }
  }

  private def buildAuditEvents(
    projects: LazyList[Project],
    orgMemberMap: Map[String, Vector[String]],
    projectOrgMap: Map[String, String],
    apiKeys: LazyList[ApiKey],
    invitations: LazyList[Invitation],
    rng: Random
  ): LazyList[AuditEvent] = {
    val apiKeysByOrg = apiKeys.groupBy(_.org_id)
    val invitationsByOrg = invitations.groupBy(_.org_id)

    projects.flatMap { proj =>
      val orgId = projectOrgMap(proj.id)
      val memberIds = orgMemberMap.getOrElse(orgId, Vector.empty)
      val orgApiKeys = apiKeysByOrg.getOrElse(orgId, Vector.empty)
      val orgInvitations = invitationsByOrg.getOrElse(orgId, Vector.empty)

      if (memberIds.isEmpty) Seq.empty
      else {
        val numEvents = 5 + rng.nextInt(496) // 5 to 500

        (1 to numEvents).map { _ =>
          val userId = memberIds(rng.nextInt(memberIds.length))
          val action = Distributions.weightedAction(rng)
          val resourceType = Distributions.weightedResourceType(rng)
          
          val resourceId = resourceType match {
            case "project" => proj.id
            case "membership" => Pools.deterministicUuid(rng) // Mocking resource id
            case "user" => userId
            case "api_key" => if (orgApiKeys.nonEmpty) orgApiKeys(rng.nextInt(orgApiKeys.length)).id else proj.id
            case "invitation" => if (orgInvitations.nonEmpty) orgInvitations(rng.nextInt(orgInvitations.length)).id else proj.id
            case _ => proj.id
          }

          val projCreatedInstant = Instant.parse(proj.created_at).toEpochMilli
          val now = Instant.parse("2024-01-01T00:00:00Z").toEpochMilli
          val range = math.max(1000L, now - projCreatedInstant)
          val tsInstant = projCreatedInstant + (rng.nextDouble() * range).toLong
          val ts = Instant.ofEpochMilli(tsInstant).toString

          AuditEvent(
            id = Pools.deterministicUuid(rng),
            org_id = orgId,
            user_id = userId,
            project_id = proj.id,
            action = action,
            resource_type = resourceType,
            resource_id = resourceId,
            ts = ts
          )
        }
      }
    }
  }
}

File: examples/canonical-faker/elixir/lib/tenant_core/pipeline.ex
defmodule TenantCore.Pipeline do
  alias TenantCore.{Distributions, Pools}

  def generate(%TenantCore.Profile{} = profile) do
    :rand.seed(:exsss, {profile.seed, 0, 0})
    
    orgs = build_orgs(profile.num_orgs)
    users = build_users(profile.num_users)
    user_ids = Enum.map(users, & &1.id)

    {memberships, org_member_map} = build_memberships(orgs, user_ids)
    {projects, project_org_map} = build_projects(orgs, org_member_map)
    audit_events = build_audit_events(projects, org_member_map, project_org_map)
    api_keys = build_api_keys(orgs, org_member_map)
    invitations = build_invitations(orgs, org_member_map)

    dataset = %{
      organizations: orgs,
      users: users,
      memberships: memberships,
      projects: projects,
      audit_events: audit_events,
      api_keys: api_keys,
      invitations: invitations
    }

    report = TenantCore.Validate.check(dataset)
    unless report.ok, do: raise("Validation failed: #{inspect(report.violations)}")
    Map.put(dataset, :report, report)
  end

  defp deterministic_uuid do
    bytes = for _ <- 1..16, into: <<>>, do: <<:rand.uniform(256) - 1>>
    <<u0::48, _::4, u1::12, _::2, u2::62>> = bytes
    <<u0::48, 4::4, u1::12, 2::2, u2::62>>
    |> Base.encode16(case: :lower)
    |> (fn <<a::8-bytes, b::4-bytes, c::4-bytes, d::4-bytes, e::12-bytes>> -> "#{a}-#{b}-#{c}-#{d}-#{e}" end).()
  end

  defp deterministic_timestamp do
    base = ~N[2020-01-01 00:00:00]
    offset = :rand.uniform(365 * 3 * 24 * 60 * 60)
    (NaiveDateTime.add(base, offset, :second) |> NaiveDateTime.to_iso8601()) <> "Z"
  end

  defp deterministic_string(len) do
    bytes = for _ <- 1..len, into: <<>>, do: <<:rand.uniform(256) - 1>>
    Base.encode16(bytes, case: :lower) |> String.slice(0, len)
  end

  defp build_orgs(count) do
    Enum.map(1..count, fn _ ->
      %TenantCore.Organization{
        id: deterministic_uuid(),
        name: Faker.Company.name(),
        slug: Faker.Internet.slug(),
        plan: Distributions.weighted_plan(nil),
        region: Distributions.weighted_region(nil),
        created_at: deterministic_timestamp(),
        owner_email: Faker.Internet.email()
      }
    end)
  end

  defp build_users(count) do
    Enum.map(1..count, fn _ ->
      locale = Distributions.weighted_locale(nil)
      %TenantCore.User{
        id: deterministic_uuid(),
        email: Faker.Internet.email(),
        full_name: Faker.Person.name(),
        locale: locale,
        timezone: Distributions.timezone_for_locale(locale),
        created_at: deterministic_timestamp()
      }
    end)
  end

  defp build_memberships(orgs, user_ids) do
    memberships =
      Enum.flat_map(orgs, fn org ->
        num_members = :rand.uniform(10) + 2
        # Enum.take_random is not purely driven by :rand for small lists? Wait, it uses :rand.
        # But to be safe we can use deterministic shuffle if take_random fails.
        # In Elixir, Enum.take_random uses :rand.
        members = Enum.take_random(user_ids, num_members)
        
        Enum.with_index(members)
        |> Enum.map(fn {uid, idx} ->
          role = if idx == 0, do: "owner", else: Distributions.weighted_role(nil)
          %TenantCore.Membership{
            id: deterministic_uuid(),
            org_id: org.id,
            user_id: uid,
            role: role,
            joined_at: org.created_at,
            invited_by: nil
          }
        end)
      end)
    {memberships, Pools.build_org_member_map(memberships)}
  end

  defp build_projects(orgs, org_member_map) do
    projects =
      Enum.flat_map(orgs, fn org ->
        members = Map.get(org_member_map, org.id, [])
        if members == [] do
          []
        else
          num_projects = :rand.uniform(5)
          Enum.map(1..num_projects//1, fn _ ->
            %TenantCore.Project{
              id: deterministic_uuid(),
              org_id: org.id,
              name: Faker.Commerce.product_name(),
              status: Distributions.weighted_project_status(nil),
              created_by: Enum.random(members),
              created_at: org.created_at
            }
          end)
        end
      end)
    {projects, Pools.build_project_org_map(projects)}
  end

  defp build_audit_events(projects, org_member_map, _project_org_map) do
    Enum.flat_map(projects, fn project ->
      members = Map.get(org_member_map, project.org_id, [])
      if members == [] do
        []
      else
        num_events = :rand.uniform(5)
        Enum.map(1..num_events//1, fn _ ->
          %TenantCore.AuditEvent{
            id: deterministic_uuid(),
            org_id: project.org_id,
            user_id: Enum.random(members),
            project_id: project.id,
            action: Distributions.weighted_action(nil),
            resource_type: Distributions.weighted_resource_type(nil),
            resource_id: deterministic_uuid(),
            ts: project.created_at
          }
        end)
      end
    end)
  end

  defp build_api_keys(orgs, org_member_map) do
    Enum.flat_map(orgs, fn org ->
      members = Map.get(org_member_map, org.id, [])
      if members == [] do
        []
      else
        num_keys = :rand.uniform(3) - 1
        if num_keys <= 0 do
          []
        else
          Enum.map(1..num_keys, fn _ ->
            %TenantCore.ApiKey{
              id: deterministic_uuid(),
              org_id: org.id,
              created_by: Enum.random(members),
              name: Faker.Lorem.word(),
              key_prefix: "tc_" <> deterministic_string(8),
              created_at: org.created_at,
              last_used_at: org.created_at
            }
          end)
        end
      end
    end)
  end

  defp build_invitations(orgs, org_member_map) do
    Enum.flat_map(orgs, fn org ->
      members = Map.get(org_member_map, org.id, [])
      if members == [] do
        []
      else
        num_invites = :rand.uniform(3) - 1
        if num_invites <= 0 do
          []
        else
          Enum.map(1..num_invites, fn _ ->
            %TenantCore.Invitation{
              id: deterministic_uuid(),
              org_id: org.id,
              invited_email: Faker.Internet.email() <> ".inv",
              role: Distributions.weighted_role(nil),
              invited_by: Enum.random(members),
              expires_at: deterministic_timestamp(),
              accepted_at: nil
            }
          end)
        end
      end
    end)
  end
end
